core/src/main/java/smile/manifold/UMAP.java

```diff
@@ -0,0 +1,846 @@
+                        .mapToDouble(neighbor -> (neighbor != null && neighbor.weight > 0 ? neighbor.weight : 0))
+                        .average().orElse(0);
+                sigmas[i] = Math.max(sigmas[i], MIN_K_DIST_SCALE * mean_ith_distances);
+            } else {
+                sigmas[i] = Math.max(sigmas[i], MIN_K_DIST_SCALE * meanDistance);
+            }
+        } // end for iteration over every data point:knns
+    }
+
+    /**
+     * Construct the membership strength data for the 1-skeleton of each local
+     * fuzzy simplicial set. This is formed as a sparse matrix where each row is
+     * a local fuzzy simplicial set, with a membership strength for the
+     * 1-simplex to each other data point.
+     * 
+     * @param knns
+     *            n nearest points for each data point
+     * @param sigmas
+     *            Smooth approximator to knn-distance from {@link UMAP#smoothKnnDist}
+     * @param rhos
+     *            Distance to nearest neighbor from {@link UMAP#smoothKnnDist}
+     * @return A fuzzy simplicial set represented as a sparse matrix. The (i, j)
+     *         entry of the matrix represents the membership strength of the
+     *         1-simplex between the ith and jth sample points.
+     */
+    private static SparseMatrix computeMembershipStrengths(Graph knns, double[] sigmas, double[] rhos) {
+
+        int size = knns.getNumVertices();
+
+        // construct original coo-format sparse matrix from knns
+        Map<Integer, Map<Integer, Double>> cooColumns = new HashMap<>();
+        IntStream.range(0, size).forEach(n -> cooColumns.put(n, new HashMap<Integer, Double>()));
+        int nonZeros = 0;
+        for (int i = 0; i < size; i++) {
+            Collection<Edge> neighbors = knns.getEdges(i);
+            if (neighbors == null || neighbors.size() <= 0) {
+                continue;
+            }
+            for (Edge edge : neighbors) {
+                if (edge == null) {
+                    continue;
+                }
+                double neighborDistance = edge.weight;
+                if (neighborDistance < 0) {
+                    continue;
+                }
+                double val = -1;
+                if (edge.v2 == i) {
+                    val = 0.0;
+                } else if (neighborDistance <= rhos[i] || sigmas[i] == 0.0) {
+                    val = 1.0;
+                } else {
+                    val = Math.exp(-(neighborDistance - rhos[i]) / (sigmas[i]));
+                }
+
+                if (val > 0) {
+                    cooColumns.get(edge.v1).put(edge.v2, val);
+                    nonZeros++;
+                }
+            }
+        }
+        SparseMatrix coo = convert2SpareMatrix(cooColumns, size, nonZeros);
+
+        // we need to re-construct the sparse matrix using probabilistic
+        // t-conorm: (a + b - a * b)
+        SparseMatrix transpose = coo.transpose();
+        SparseMatrix product = coo.aat();
```

Comment:
This is wrong. `aat` is matrix multiplication. The paper uses pointwise product.